feat: permanently save logs/events

alaibe opened the pull request:
![circular-spaceship-18218-2560x1440](https://user-images.githubusercontent.com/491074/48343731-0f74c480-e66b-11e8-824a-3005d7bb0a15.jpg)

PascalPrecht commented:
Can we update the scope to `modules/blockchain_connector` here?

alaibe commented:
It touches multiple scopes

PascalPrecht commented:
Yes good point. Ignore my comment! :)

michaelsbradleyjr commented on src/lib/core/fs.js:
```diff
@@ -141,6 +141,10 @@ function existsSync() {
 }
 
+function ensureFileSync() {
+  return restrictPath(fs.ensureFileSync, fs.existsSync, 1, arguments);
```
```suggestion
  return restrictPath(fs.ensureFileSync, fs.ensureFileSync, 1, arguments);
```
^ all the others have the first and second arguments being the same, is this one a special case?

alaibe replied:
Updated

michaelsbradleyjr commented on src/lib/modules/blockchain_connector/index.js:
```diff
@@ -699,6 +702,26 @@ class BlockchainConnector {
+  _saveEvent(event) {
+    const data = this._readEvents();
+    data[new Date().getTime()] = event;
+    fs.writeJSONSync(this.logFile, data);
```
If this will result in a lot of sync fs access, should we use async (`writeJson`) instead of sync? We'd need to save events to a queue, drain/wait, then loop back. I don't want things to be over-complicated but I'd hate to accidentally started bogging down embark.

alaibe replied:
I feel it would indeed be over-complicated.
Also those actions are done when embark is doing nothing but waiting.

The code-generator perform a lot more file system action without using such technique and performance is more than ok

jrainville commented on src/lib/modules/console_listener/index.js:
```diff
@@ -121,10 +124,30 @@ class ConsoleListener {
+  _saveLog(log) {
+    const data = this._readLogs();
+    data[new Date().getTime()] = log;
+    fs.writeJSONSync(this.logFile, data);
```
Same thing for async here.

alaibe commented:
@jrainville exactly, keep the event between stop and start and also low memory.
The problem of switching to not async is concurrent write.

So if we switch to async I feel we should do the queue too.

jrainville commented:
Ah ok, I understand your point with the `async`. But I feel like it would be cleaner and better to do it anyway. Especially with the `async` library, it's really simple. 
Just use a cargo to "queue" the write jobs for you, then you write to the file in one go: http://caolan.github.io/async/docs.html#cargo

alaibe commented:
Fair enough, I ll make the changes

michaelsbradleyjr commented:
async cargo may be the way to go, but I was wondering whether this might be a good use case for [`fs.createWriteStream`](https://nodejs.org/docs/latest-v8.x/api/fs.html#fs_fs_createwritestream_path_options).

alaibe commented:
@michaelsbradleyjr I had thought about `fs.createWriteStream` and I wonder what would be the algorithm you have in mind?
1 - close the stream each time you finish the write?
I wasn't use this solution is really a win
2 - keep it open while embark is running
My problem with this one is that that I need to have a json format.
3 - something else?

IMO createWriteStream would be an alternative if just using `append`.
Anything better?

alaibe commented:
@jrainville @michaelsbradleyjr I have added the queue (works better than cargo)
If you have an idea for createWriteStream let me know an I will change.

I also added a commit for `start:embark`

jrainville commented on src/lib/modules/blockchain_connector/index.js:
```diff
@@ -24,6 +25,16 @@ class BlockchainConnector {
+        }
+        callback();
+      });
+    }, 1);
```
Problem with the queue is you can only have one job done at a time. So if 5 events are triggered in fast succession, you will have to write 5 times one after the other. (including 5 reads)

Whereas, with the cargo, you will receive an array of tasks (the 4/5 json datas) that you can iterate over before actually writing. Example:
```
    this.writeLogFile = async.cargo((tasks, callback) => {
      const data = this._readEvents();

      tasks.forEach(task => {
          data[new Date().getTime()] = task;
      });

      fs.writeJson(this.logFile, data, err => {
        if (err) {
          console.error(err);
        }
        callback();
      });
    }); // Put here no limit, as we can handle all tasks
```

alaibe replied:
TIL 💯 will make the change

michaelsbradleyjr replied:
I have a createWriteStream variant almost ready too, after the exchange of comments yesterday I decided to give it a shot.

michaelsbradleyjr commented:
@alaibe, @jrainville: here's my attempt at an implementation with a Writable stream:

https://github.com/embark-framework/embark/compare/feature/permanent-save-log-event...feature/permanent-save-log-event-WriteStream

A few things to note:

* Per my discussion with @alaibe, there seems to be a problem with event logging on this PR branch, so my branch only has changes for the console listener.
* Node's [docs for the `write` method]((https://nodejs.org/dist/latest-v10.x/docs/api/stream.html#stream_writable_write_chunk_encoding_callback)) indicate a stream will buffer if it's not draining; so, I'm not sure if implementing n queue with an array was really necessary or provides any benefits for us, but that's just how I did it.
* Node's [docs for `fs.createWriteStream`](https://nodejs.org/dist/latest-v10.x/docs/api/fs.html#fs_fs_createwritestream_path_options) indicate that `autoClose` is enabled by default, and will trigger if there's an error, closing the file descriptor. My code doesn't listen for the `error` event, but maybe that should be added and on error we could create a fresh file, not sure what's best.
* I'm not super happy with the loops of 1 ms timers (to avoid reader/writer collision). It would be better to have `logFileWriting` and `logFileReading` implemented as observables and await on a promise that resolves when the observable has an event... something like that. But that's something to improve in the future, I think.
* If we're streaming the writes, maybe we can stream the reads too, piping them to the `res` object for more efficient delivery?
* The log file isn't really valid JSON or [NDJSON](http://ndjson.org/). Should we give it a different file extension, maybe just `.log`?

michaelsbradleyjr commented:
I made some important changes in `_saveLog` so that the queue is used much more effectively, i.e. pending writes don't pile up looped timers, they just write to the queue. In fact, I was able to get rid of the looped time `_saveLog`, it's only needed in `_readLogs` now.

alaibe commented:
@jrainville PR updated with cargo
@michaelsbradleyjr I had a good look at your implementation.
IMO it is way to much complexity for the usage. Also with the file queue you have re-implemented exactly what cargo does. The boolean: `this.logFileReading` won't work you need a lock strategy here in order to be really efficient I think.

I also feel that json makes it more easy, it was the main reason why I didn't go for Stream in the first place.

michaelsbradleyjr commented:
@alaibe The queue I implemented is definitely similar. I think going with a simple, straightforward solution like cargo is worthwhile, but I still have some questions.

* `_readLogs` is implemented with `readFileSync`, and every call of `_saveLog()` results in a call of `_readLogs()`. Is the idea with `readFileSync` to avoid reader-writer collision?

If so, and if it's the only approach that's practical for low complexity, then maybe it's just as well that we use `writeJsonSync`, since every call of `_readLogs()` will block the process anyway.

* If the time stamps are tossed in the end, and since `Object.values()` doesn't guarantee key ordering (though maybe it works in practice), how about encoding as an array of objects (`JSON.stringify('[{},{},{}]')`)?

New entries would simply be pushed onto the end of the array.

* With every call of `_saveLog()` we're reading and writing the entire log. It can definitely work at a small scale, but if the log sticks around for awhile and/or if the logging rate is high for some reason, won't that become a bottleneck? 

Perhaps at some point we should investigate use of [fs-ndjson](https://www.npmjs.com/package/fs-ndjson) or similar.

* For my own understanding, can you explain why `this.logFileReading` doesn't work?

I experimented in VS Code's debugger and it did seem to work (reading blocking writing, and vice versa). It's a given that my simple locking strategy can't work across multiple instances of `ConsoleListener` accessing the same file, and it provides no protection re: other code/processes that might read/write to the same file.

michaelsbradleyjr commented:
I have learned that in the ES6 specification, object property traversal became well defined:

http://2ality.com/2015/10/property-traversal-order-es6.html

So please that aspect of my question/concern about using an array vs. an object.

alaibe commented:
@michaelsbradleyjr 

> _readLogs is implemented with readFileSync, and every call of _saveLog() results in a call of _readLogs(). Is the idea with readFileSync to avoid reader-writer collision?

Definitively

>If so, and if it's the only approach that's practical for low complexity, then maybe it's just as well that we use writeJsonSync, since every call of _readLogs() will block the process anyway.

The only advantage of cargo is that if you have 10 simultaneous request, it will result in 1 read/1write
because it will pack the request together. See this link for an explanation of cargo:
https://camo.githubusercontent.com/f4810e00e1c5f5f8addbe3e9f49064fd5d102699/68747470733a2f2f662e636c6f75642e6769746875622e636f6d2f6173736574732f313637363837312f36383130312f38346339323036362d356632392d313165322d383134662d3964336430323431336266642e676966

> If the time stamps are tossed in the end, and since Object.values() doesn't guarantee key ordering (though maybe it works in practice), how about encoding as an array of objects (JSON.stringify('[{},{},{}]'))?
New entries would simply be pushed onto the end of the array.

You reply to this one yourself ;)

>With every call of _saveLog() we're reading and writing the entire log. It can definitely work at a small scale, but if the log sticks around for awhile and/or if the logging rate is high for some reason, won't that become a bottleneck?
Perhaps at some point we should investigate use of fs-ndjson or similar.

I definitively agree that it won't work very well if we start to have Gb of data.
However those logs are save only in development when there is tx. I doubt we will have someone with Gb of data any time soon.
Also if you look at the read, it reads the full file, this is definitively not scalable.
We would need to introduce pagination/cursor IMO.
I didn't do it because the way we handle the storage is temporary. If we introduce sqlite for example, most of the problems are solved.

>For my own understanding, can you explain why this.logFileReading doesn't work?
I experimented in VS Code's debugger and it did seem to work (reading blocking writing, and vice versa). It's a given that my simple locking strategy can't work across multiple instances of ConsoleListener accessing the same file, and that it provides no protection re: other code/processes that might read/write to the same file.

Here we are talking about race condition. I have personally never being able to reproduce a race condition with a debugger. The reason for that is that usually those errors happens at micro second speeds and when you debug you reduce the speed.
Also note that with one user (the developer), the race condition is unlikely to happen.
Here is why I think the locking strategy won't work:
1) save_log is called and go to line 150
2) read_log is called and go to line 176
From there you can see that whatever happens next you can have a read and write simultaneously
